pallets/capacity/src/lib.rs

```diff
@@ -1036,6 +1046,26 @@ impl<T: Config> Pallet<T> {
 		Self::set_capacity_for(*to_msa, capacity_details);
 		Ok(())
 	}
+
+	/// updates or inserts a new boost history record for current_era.   Pass 'add' = true for an increase (provider_boost),
+	/// pass 'false' for a decrease (unstake)
+	pub(crate) fn upsert_boost_history(
+		account: &T::AccountId,
+		current_era: T::RewardEra,
+		boost_amount: BalanceOf<T>,
+		add: bool,
+	) -> Result<(), DispatchError> {
+		let mut boost_history = Self::get_staking_history_for(account).unwrap_or_default();
+
+		let upsert_result = if add {
+			boost_history.add_era_balance(&current_era, &boost_amount)
+		} else {
+			boost_history.subtract_era_balance(&current_era, &boost_amount)
+		};
+		ensure!(upsert_result.is_some(), Error::<T>::EraOutOfRange);
+		ProviderBoostHistories::<T>::set(account, Some(boost_history));
```

Comment:
One thing - unstaking from a boosted target doesn't need a separate extrinsic, since at some point it was decided not to support staking to the same target in two different ways.  Once the account is found it will know what type of staking was done.

Yes, to your last point.  If you did a double key map then that would make it harder to put limits on staking history when looking for the oldest record.  You would still need to iterate over all the ones in there on a new `provider_boost` to ensure the history limit is respected, so it doesn't save you anything.

The `BoundedBTreeMap` is specifically supposed to be a balance of memory use and performance between a BTreeMap and a BoundedVec, and it gives you convenience methods like `first_key_value` and `last_key_value` if you need/want meaningfully sortable keys.

So here is another example:
1.  Say the history limit is 10, and I stake 100 in era 10.
2.  Over the next 20 eras I stake 10 another 9 times. I have filled up my history limit, and I still have the first record.
3.  I go to claim in era 31, but the rewards for eras from 11-19 have expired.  I get only the rewards from eras 20-29 (maybe there is an off-by-one error, but you get the picture)